### PR TITLE
Device Linking and better linking ui

### PR DESCRIPTION
**Pull request by juliangiebel**
## About the PR

This basically moves machine linking on top of of device networking, allowing it to use device networking to invoke ports and send data along.
Functionality wise device linking is a drop in replacement for machine linking except some renamed stuff.

Most notably transmitters are called sources and receivers are called sinks now.

Using device networking is optional and if either the source or the sink doesn't have a `DeviceNetworkComponent` it just raises the `SignalReceivedEvent` without using networking or sending data along.

When using device networking, links can use connection checks like the `WirelessNetworkConnection` or the `StationBoundNetworkConnection` (form #7542) and work alongside normal device networking.

The UI for linking also got updated and now uses bézier curves to display links instead of straight lines.

Linking is done with the Network Configurator, which also does configuring device lists.
The Network Configurator has two modes: **List** and **Link**.
You can switch between modes by pressing the *alt interact hand* key bind. Which is *alt+z* or *alt+y* by default.

Linking devices together using device linking is done by switching the Network Configurator to **Link** mode and clicking on a source device and a sink device. Just link machine linking.

Device linking also supports machine linkings alt interaction and auto linking.

**Media**

### The new linking ui
image

### The item status added for the network configurator
image

- [x] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

**Changelog**

:cl:
- add: The network configurator. A tool used to link devices together.
- tweak: Configuring device lists is now done using the network configurator instead of the multitool.

### CODE REVIEW

**Comment by juliangiebel** [comment]
I'm almost done with this PR but I still need to do some things:
- [ ] Write proper doc comments
- [ ] Write a guidebook entry
- [ ] Add ways to obtain the network configurator

**Comment by juliangiebel** [comment]
The yaml linter says `CC-0` is not a proper license

**Comment by 0x6273** [comment]
> The yaml linter says `CC-0` is not a proper license

I think you need `CC0-1.0`

**Comment by ike709** [comment]
> The yaml linter says CC-0 is not a proper license

Licenses must be in the format of [SPDX Identifiers](https://spdx.org/licenses/)

**Comment by juliangiebel** [comment]
This is probably ready for review now :sweat_drops:

**Comment by juliangiebel** [comment]
Links done with machine linking already present on maps won't break because device linking translates them to device linking links on map init but mappers should still update their maps when this gets merged.

**Comment by Visne** [comment]
I've fixed conflicts for you

**Comment by juliangiebel** [comment]
This PR is getting unrelated test failures from the PR that renames the janicart.

**Comment by Visne** [comment]
Should be fixed after tests finish running

File: Resources/Locale/en-US/devices/network-configurator.ftl [hunk 1]
```diff
@@ -24,3 +33,13 @@ network-configurator-tooltip-edit = Edit targets device list
...
+
+# examine
+network-configurator-examine-mode-link = [color=red]Link[/color]
+network-configurator-examine-mode-list = [color=green]List[/color]
+network-configurator-examine-current-mode = Current mode: {$mode}
+network-configurator-examine-switch-modes = Press {$key} to switch modes
+
+# item status
+network-configurator-item-status-label = Current mode: {$mode}
+{$keybinding} to switch mode
```
Comment (AJCM-git) [inline Resources/Locale/en-US/devices/network-configurator.ftl:45]: oop

File: Resources/Locale/en-US/machine-linking/port-selector.ftl [hunk 2]
```diff
@@ -1,3 +1,5 @@
 signal-port-selector-menu-title = Port Selector
 signal-port-selector-menu-clear = Clear
 signal-port-selector-menu-link-defaults = Link defaults
+signal-port-selector-help=Select the ports you want to link up
+signal-port-selector-menu-done=Done
```
Comment (AJCM-git) [inline Resources/Locale/en-US/machine-linking/port-selector.ftl:5]: space (ie `help = Select`)

File: Resources/Textures/Objects/Tools/network_configurator.rsi/meta.json [hunk 3]
```diff
@@ -0,0 +1,32 @@
+{
+  "version": 1,
+  "license": "CC-BY-SA-3.0",
+  "copyright": "Julian Giebel <juliangiebel@live.de>",
```
Comment (AJCM-git) [inline Resources/Textures/Objects/Tools/network_configurator.rsi/meta.json:4]:
```suggestion
  "copyright": "Made by Julian Giebel <juliangiebel@live.de> for ss14",
```

**Review by AJCM-git — CHANGES_REQUESTED** [review]

**Review by AJCM-git — CHANGES_REQUESTED** [review]
Its kinda late so tell me if something i rambled about here doesnt make sense/isnt possible/is out of scope:
- An item mapper for the network configurator in the belt would be nice (if there is one already it isnt working)
- Add guidehelpcomponent mapped to networking/net config guides to entities related to those (extra points if you a book with those two guides) 
- Update the localized messages that say "Link X with a `multitool`"
- Network configurator should start on Link mode by default
- Make net config link BUI open centered to the right, not in the middle (the small one does, the big one with copy, show, etc buttons doesnt)
- The "Show" overlay of the list BUI stays active even if you dont have the net configurator in your hand or its BUI open
- Try making that overlay always active with the engineering googles (maybe behind an action?) if its not too out-of-scope
- Make "link defaults" work with common machines (like airlocks and buttons, cloning pod/console, etc)
- The link net configurator BUI doesnt update when you have it open and click on a device again to stop linking
- The list mode Net configurator BUI is very confusing on first use, you can open two of them that are different from each other if you click on an airlock and use the configurator and then click on a cloning computer. I know the small one is the one you have "locally" on your net configurator and the big one the one the computer has, but they dont make this difference known which leads to confusion, also the big one probably shouldnt even open on devices that dont support lists like the cloning console

**Comment by juliangiebel** [comment]
> Its kinda late so tell me if something i rambled about here doesnt make sense/isnt possible/is out of scope:
> ...

Thanks for the review 👍 

Most of the points you mentioned aren't out of scope.
Changing the overlay is a little bit out of scope though and I want to improve it anyway so I will do that together with the improvement I plan on doing in another PR.

A few machines that used machine linking and got ported over to device linking have the device list component for some reason, even though they don't use device lists. I'm just gonna remove that component from them.

> An item mapper for the network configurator in the belt would be nice (if there is one already it isnt working)

What do you mean by that?

---

I'm also not 100% sure if I like the mode switching or if I should just make it open the right ui by default. Most things wouldn't use linking and lists at the same time and if they do I can just have the list ui open using alt interact.
What are your thoughts on switching modes vs just opening the correct UI and using alt interact for edge cases?

File: Resources/Locale/en-US/devices/network-configurator.ftl (hunk 1 above)
Comment (juliangiebel) [reply Resources/Locale/en-US/devices/network-configurator.ftl:45]: That's actually valid ftl
The linebreak carries over into the ui where that line is displayed

**Comment by AJCM-git** [comment]
> What do you mean by that?

the belt prototype has a item mapper component thats used for the tools appearance in the belt sprite

> What are your thoughts on switching modes vs just opening the correct UI and using alt interact for edge cases?

I think opening the correct UI and still having the option to switch modes would be best, with a message in the ui or a popup telling you that the device doesnt support linking/listings and maybe a "switch to list /link" in the UI, having a way to differentiate the target list UI from the network configurator list would also help (maybe separate some of their behavior because if you close the target one it closes the configurator one)

**Comment by juliangiebel** [comment]
> Make net config link BUI open centered to the right, not in the middle (the small one does, the big one with copy, show, etc buttons doesnt)
@AJCM-git It opens centered to not conflict with the small one (the saved devices ui)

**Comment by AJCM-git** [comment]
> It opens centered to not conflict with the small one (the saved devices ui)

make it open centered to the left then

**Review by ElectroJr — COMMENTED** [review]
I might be a bit busy up until the weekend but I'll review & can help resolve conflicts sometime in the next 7 days.

File: Content.Shared/DeviceLinking/DeviceLinkSinkComponent.cs [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+    /// <summary>
+    /// The ports this sink has
+    /// </summary>
+    [DataField("ports", customTypeSerializer: typeof(PrototypeIdHashSetSerializer<SinkPortPrototype>))]
+    public HashSet<string>? Ports;
+
+    /// <summary>
+    /// Used for removing a sink from all linked sources when it gets removed
+    /// </summary>
+    [DataField("links")]
```
[the remark this answers is not in the document]
Comment (ElectroJr) [reply Content.Shared/DeviceLinking/DeviceLinkSinkComponent.cs:18]: this really shouldn't be written to via VV, and data-fields already default to read-only VV

File: Content.Shared/DeviceLinking/Events/LinkAttemptEvent.cs [hunk 5]
```diff
@@ -0,0 +1,20 @@
+namespace Content.Shared.DeviceLinking.Events
+{
+    public sealed class LinkAttemptEvent : CancellableEntityEventArgs
```
[the remark this answers is not in the document]
Comment (ElectroJr) [reply Content.Shared/DeviceLinking/Events/LinkAttemptEvent.cs:3]: I changed them to global name space, but if you meant the ref event one, its not really necessary as they're called very infrequently and then you can't just use `: CancellableEntityEventArgs`

**Review by ElectroJr — COMMENTED** [review]
AFAICT RSI diff test error is just because it can't cope with file paths with spaces, and this PR modifies files in `Resources/Server Info`

**Pull request merged**